test/extensions/io_socket/user_space/io_handle_impl_test.cc

```diff
@@ -1056,6 +1056,24 @@ TEST_F(IoHandleImplTest, ActivateEvent) {
   ASSERT_TRUE(schedulable_cb_->enabled());
 }
 
+// This is a compatibility test for Envoy Connection. When a connection is destroyed, the Envoy
+// connection may close the underlying handle but not destroy that io handle. Meanwhile, the
+// Connection object does not expect any further event be invoked because the connection in destroy
+// pending state can not support read/write.
+TEST_F(IoHandleImplTest, EventCallbackIsNotInvokedIfHandleIsClosed) {
+  testing::MockFunction<void()> check;
+  schedulable_cb_ = new NiceMock<Event::MockSchedulableCallback>(&dispatcher_);
+  io_handle_->initializeFileEvent(
+      dispatcher_, [&, handle = io_handle_.get()](uint32_t) { check.Call(); },
+      Event::FileTriggerType::Edge, Event::FileReadyType::Read);
+  EXPECT_FALSE(schedulable_cb_->enabled());
+  io_handle_->activateFileEvents(Event::FileReadyType::Read);
+  EXPECT_TRUE(schedulable_cb_->enabled());
+
+  EXPECT_CALL(check, Call()).Times(0);
+  io_handle_->close();
```

Comment:
I think that this test will succeed without your changes to io_handle_impl.cc

Do you need to run the dispatcher loop here?

Reply:
Make sense I will update

Reply:
The schedulable is not driven by the dispatcher. It's a mock schedulable. 
 The challenge is that before or after the fix, the life time of  schedulable is different. I am adding a onDestroy at the mock schedulable.

To write a failing in past test case via the onDestroy is straighforward, however, that ondestroy breaks other test cases and I am fixing

Reply:
Eventually I add another destroy callback parameter in the MockSchedualbleCallback. Adding ON_CALL( onDestroy) break the http2 test because of the InSequence usage and I didn't find a easy way to add partial order sequence